pkg/storage/stores/shipper/downloads/table.go

```diff
@@ -0,0 +1,400 @@
+package downloads
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/chunk"
+	"github.com/cortexproject/cortex/pkg/chunk/local"
+	chunk_util "github.com/cortexproject/cortex/pkg/chunk/util"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"go.etcd.io/bbolt"
+)
+
+// timeout for downloading initial files for a table to avoid leaking resources by allowing it to take all the time.
+const downloadTimeout = 5 * time.Minute
+
+type BoltDBIndexClient interface {
+	QueryDB(ctx context.Context, db *bbolt.DB, query chunk.IndexQuery, callback func(chunk.IndexQuery, chunk.ReadBatch) (shouldContinue bool)) error
+}
+
+type downloadedFile struct {
+	mtime  time.Time
+	boltdb *bbolt.DB
+}
+
+// Table is a collection of multiple files created for a same table by various ingesters.
+// All the public methods are concurrency safe and take care of mutexes to avoid any data race.
+type Table struct {
+	name              string
+	cacheLocation     string
+	metrics           *metrics
+	storageClient     chunk.ObjectClient
```

Comment:
You use only List and GetObject, you could lower the dependency requirement by creating your own interface declaring what you want to use.

Reply:
Like `BoltDBIndexClient` I also like the fact that it document the package.